### ISSUE TITLE
Spurious SC2004 with multiple files (0.9.0 regression)

### ISSUE THREAD

**Issue by CyberShadow** [koalaman/shellcheck#2851]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2004
- My shellcheck version (`shellcheck --version` or "online"): 0.9.0
- [X] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

`main.sh`:
```sh
#!/bin/bash
source ./a.sh
source ./b.sh
```
`a.sh`:
```sh
declare -A file_content_filters
```
`b.sh`:
```sh
pattern=key
file_content_filters[$pattern]=value

printf '%q\n' "${!file_content_filters[@]}"
```

#### Here's what shellcheck currently says:

```console
$ shellcheck -x -a ./main.sh

In ./b.sh line 2:
file_content_filters[$pattern]=value
                     ^------^ SC2004 (style): $/${} is unnecessary on arithmetic variables.

For more information:
  https://www.shellcheck.net/wiki/SC2004 -- $/${} is unnecessary on arithmeti...
```

#### Here's what I wanted or expected to see:

No warning, as with a single file, and as with ShellCheck 8.x
